uips/uip-4.md

```diff
@@ -0,0 +1,102 @@
+# UIP: Spend Backreferences
+
+| uip | 4 |
+| - | - |
+| title | Spend Backreferences |
+| description | Spend Backreferences enable improved sync performance |
+| author | Jennifer Helsby ([@redshiftzero](https://github.com/redshiftzero)), Henry de Valence ([@hdevalence](https://github.com/hdevalence)), Lúcás Meier ([@cronokirby](https://github.com/cronokirby))|
+| discussions-to | <https://forum.penumbra.zone/t/uip-spend-backreferences/110> |
+| status | Draft |
+| type | Standards Track |
+| consensus | Yes |
+| created | 2024-11-06 |
+
+## Abstract
+
+This specification introduces a method to improve Penumbra sync speeds by adding additional data that can be used by DAGSync clients. `Spend` actions will contain a new `backref_commitment` field, allowing clients to traverse their transaction graph backwards and quickly recover their entire transaction history.
+
+## Motivation
+
+DAGSync is a graph-aware fast syncing algorithm. A client, upon detecting a single transaction involving them, can check that outputs visible to them are spent or not. If the output is unspent, then they have identified a live note they can potentially spend in the future, else if the output is unspent, they can continue the process forwards in the transaction graph, until they reach unspent notes.
+
+The design of Penumbra currently does not allow traversal backwards through the transaction graph, only forwards. A `Spend` intentionally does not reveal the note being spent, only the nullifier that is revealed. By including on the `Spend` an encrypted reference back to the note commitment being spent, such that only the note owner can view it, we enable DAGSync clients to efficienctly reconstruct the transaction history both backwards and forwards.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+### Modification to `SpendBody`
+
+The `Spend` action will be augmented with an additional field `backref_commitment` on the `SpendBody`:
+
+```
+message SpendBody {
+    // A commitment to the value of the input note.
+    penumbra.core.asset.v1.BalanceCommitment balance_commitment = 1;
+    // The nullifier of the input note.
+    penumbra.core.component.sct.v1.Nullifier nullifier = 6;
+    // The randomized validating key for the spend authorization signature.
+    penumbra.crypto.decaf377_rdsa.v1.SpendVerificationKey rk = 4;
+    // NEW: An encryption of the commitment of the input note to the sender's OVK.
+    optional bytes backref_commitment = 7;
+}
+```
+
+Clients MAY populate the `backref_commitment` field with the encrypted note commitment corresponding to the note they are spending. Initially the field will be optional, allowing for a phased adoption period such that clients have time to implement support for `Spend` backreferences.
```

Comment:
Since all proto fields are optional, I think we should be more precise about the validation rules for the proto messages.

Proposal:

- Use `bytes encrypted_backref = 7;` (or whatever name we pick), no `optional`;
- Transaction parsing consensus rules require that EITHER:
  - `len(encrypted_backref) = 0`
  - `len(encrypted_backref) = 48` (or whatever the fixed size ciphertext ends up being)

Reply:
I think we should have these rules enforced, but also interpret the 0 length bytes field as being the ciphertext `[00, 00, ...]`, so that that processing code does not have to deal with the ciphertext being optional, and just have one codepath for if encryption fails

Reply:
I guess that would mess up the effect hash computation though